File: R/predict.r
#' Run prediction on cfg$las_path and write output LAS
#' @param cfg list from vegseg_config()
#' @param mode "overwrite" or "extra"
#' @param setup_env logical; FALSE if you already have Python env
#' @return output LAS path (character)
#' @export
vegseg_predict <- function(cfg, mode = c("overwrite", "extra"), setup_env = FALSE) {
  mode <- match.arg(mode)
  stopifnot(is.list(cfg))
  
  py_model  <- reticulate::import("vegseg.model",  delay_load = FALSE)
  py_infer  <- reticulate::import("vegseg.infer",  delay_load = FALSE)
  torch     <- reticulate::import("torch",         delay_load = FALSE)
  
  device <- if (is.null(cfg$device)) {
    if (torch$cuda$is_available()) "cuda" else "cpu"
  } else cfg$device
  
  message(">> Loading model weights: ", cfg$model_path)
  # Build the model skeleton to load weights
  num_classes <- if (!is.null(cfg$num_classes)) as.integer(cfg$num_classes) else 3L
  mdl <- py_model$HeightAwarePointNetTiny(in_ch = 4L, num_classes = num_classes, k = 16L,
                                          z_idx = 2L, hag_idx = 3L, use_height_prior = TRUE)
  mdl <- mdl$to(device)
  state_dict <- torch$load(cfg$model_path, map_location = device)
  mdl$load_state_dict(state_dict)
  mdl$eval()
  
  message(">> Running inference on ", cfg$las_path, " ...")
  y_pred <- py_infer$infer_on_las_path(
    model          = mdl,
    las_path       = cfg$las_path,
    BLOCK_SIZE     = as.numeric(cfg$block_size),
    STRIDE         = as.numeric(cfg$stride),
    SAMPLE_N       = as.integer(cfg$sample_n),
    REPEAT_PER_TILE= as.integer(cfg$repeat_per_tile),
    MIN_PTS_TILE   = as.integer(cfg$min_pts_tile),
    CELL_SIZE      = as.numeric(cfg$cell_size),
    QUANTILE       = as.numeric(cfg$quantile),
    DEVICE         = device
  )
  
  # decide output path
  out_dir <- cfg$out_pred_dir
  if (is.null(out_dir) || !nzchar(out_dir)) out_dir <- "data/output_predictions"
  if (!dir.exists(out_dir)) dir.create(out_dir, recursive = TRUE)
  out_las <- file.path(out_dir, "trees_predicted.las")
  
  message(">> Writing predictions to: ", out_las, " (mode=", mode, ")")
  py_infer$write_predictions_to_las(
    in_las  = cfg$las_path,
    out_las = out_las,
    y_pred  = y_pred,
    mode    = mode
  )
  
  invisible(out_las)
}

File: src/predict..py
from .config_bridge import apply_cfg
from . import infer, model as mdl
import torch

def run_predict(py_cfg, new_in_las, new_out_las, mode="overwrite"):
    class C: pass
    cfg = apply_cfg(py_cfg, C)
    device = cfg.device or ("cuda" if torch.cuda.is_available() else "cpu")

    print("Loading model:", cfg.model_path)
    m = mdl.HeightAwarePointNetTiny(in_ch=4, num_classes=3, k=16,
                                    z_idx=2, hag_idx=3, use_height_prior=True).to(device)
    state = torch.load(cfg.model_path, map_location=device)
    m.load_state_dict(state); m.eval()

    print(f"Inference on {new_in_las}...")
    y_pred = infer.infer_on_las_path(
        m, new_in_las,
        BLOCK_SIZE=cfg.block_size, STRIDE=cfg.stride,
        SAMPLE_N=cfg.sample_n, REPEAT_PER_TILE=cfg.repeat_per_tile,
        MIN_PTS_TILE=cfg.min_pts_tile,
        CELL_SIZE=cfg.cell_size, QUANTILE=cfg.quantile,
        DEVICE=device
    )
    infer.write_predictions_to_las(new_in_las, new_out_las, y_pred, mode=mode)
    print("Wrote predictions to:", new_out_las)
    return new_out_las
